24.md

```diff
@@ -0,0 +1,479 @@
+
+Gift wraps alone are an effective tool to hide information from the public. However, the inner event can be easily re-broadcasted, breaking the privacy of the sender and leaking its metadata (e.g. author, event kind, dates, tags, etc). The GiftWrap MUST wrap a Sealed Gossip Event to make sure the receiver cannot break the privacy guarantees when re-broadcasting it. 
+
+```js
+{
+  "id": "<The usual hash>",
+  "pubkey": "<Random, one-time-use PubKey>",
+  "content": "<XCHACHA-Encrypted Kind 13>",
+  "kind": 1059,
+  "created_at": 1686840217,
+  "tags": [
+    [ "p", "<Receiver>" ]
+  ],
+  "sig": "<Random, one-time-use PubKey Signature>"
+}
+```
+
+A `p` tag for the recipient MUST be added to the `tags` field of a GiftWrap. This is how clients will download incoming "things". Other tags should not be added to this event. 
+
+## 4. Encrypting with XChaCha (former NIP-44)
+
+Encryption is made with the SHA-256 hash of the shared secret between a private and a public key. 
+
+First, serialize the event to JSON. Then acquire the shared secret, a 24-byte nonce, and encrypt it with a XChaCha cipher (XChaCha is different than ChaCha). Convert both the nonce and the resulting byte array to Base64 and output an object with the following fields: 
+
+```js
+{
+  "ciphertext": "<Base64-encoded ByteArray>"
+  "nonce": "<Base64-encoded Nonce>"
+  "v": <Version Code>  
+}
+```
+
+Field `v` determines the algorithm version to be used. Current versions are: 
+ - `0`: Reserved
+ - `1`: XChaCha20
+ 
+The stringified version of this object goes into the `.content` of the GiftWrap and Sealed Gossip events.  
+ 
+## 5. Private Message Algorithm
+
+To send a private message, create a kind `Kind:14` with the message in its `.content` and a p-Tag to the receiver. Other tags are encouraged to be added, such as `content-sensitivity` if the content is NSFW and `subject` to create a name to the conversation. 
+
+Once `Kind:14` is created, remove the signature and serialize as JSON. Encrypt the serialized `Kind:14` to the receiver's public key and seal it with a `kind:13` event, signed by the sender's main key. 
+
+For each receiver: 
+
+1. Create a new key, serialize the seal, and encrypt it into a `kind:1059` with a tag to the receiver, signed by the new random key. 
+2. Broadcast the `kind:1059` event to the relays of the receiver. 
+
+For the sender: 
+
+1. Create a new key, serialize the seal, and encrypt it into a `kind:1059` with a tag to the sender, signed by the new random key. 
+2. Broadcast the `kind:1059` event to the relays of the sender. 
+
+## 6. Private Group Chat Algorithm
+
+Private Group chats are `Kind:14`s for more than one `p` tag. There is no channel definition event. The set of authors + recipient keys defines a group.
```

Comment:
So should the author tag himself?

Reply:
I merge the pubkeys in the p list with the author's pubkey so that it doesn't matter what the signing client does. :)

Reply:
We need to figure out how to remove `mentions` from being part of the group.

Reply:
I remove the author's pubkey so it doesn't matter. I guess both approaches are compatible. And yes, that's a very good point. Maybe use "cc" tags instead? They don't need to be searchable.

Reply:
Either that or we go with a "mention" marker like on Kind1s.

Reply:
That would work, but would be sort of error prone. The last thing you want is the person you're gossiping about to be invited to the conversation 😅